packages/server/modules/shared/authz.ts

```diff
@@ -0,0 +1,293 @@
+import { Scopes } from '@/modules/core/helpers/mainConstants'
+import { getRoles } from '@/modules/shared'
+import { getStream } from '@/modules/core/services/streams'
+
+import {
+  BaseError,
+  ForbiddenError as FE,
+  UnauthorizedError as UE,
+  ContextError as CE,
+  BadRequestError as BRE
+} from '@/modules/shared/errors'
+// import { getbAllRoles } from '../core/services/generic'
+
+interface AuthResult {
+  authorized: boolean
+}
+
+interface AuthFailedResult extends AuthResult {
+  authorized: false
+  error: BaseError | null
+  fatal?: boolean
+}
+
+interface Stream {
+  role?: StreamRoles
+  isPublic: boolean
+  allowPublicComments: boolean
+}
+
+interface AuthContext {
+  auth: boolean
+  role?: ServerRoles
+  stream?: Stream
+  userId?: string
+  scopes?: string[]
+}
+
+interface AuthParams {
+  streamId?: string
+}
+
+interface AuthData {
+  context: AuthContext
+  authResult: AuthResult
+  params?: AuthParams
+}
+
+interface AuthFailedData extends AuthData {
+  authResult: AuthFailedResult
+}
+
+export const authFailed = (
+  context: AuthContext,
+  error: BaseError | null,
+  fatal = false
+): AuthFailedData => ({
+  context,
+  authResult: { authorized: false, error, fatal }
+})
+
+export const authSuccess = (context: AuthContext): AuthData => ({
+  context,
+  authResult: { authorized: true }
+})
+
+enum ServerRoles {
+  Admin = 'server:admin',
+  User = 'server:user',
+  ArchivedUser = 'server:archived-user'
+}
+
+enum StreamRoles {
+  Owner = 'stream:owner',
+  Contributor = 'stream:contributor',
+  Reviewer = 'stream:reviewer'
+}
+type AvailableRoles = ServerRoles | StreamRoles
```

Comment:
Not great that ServerRoles & StreamRoles essentially duplicate everything that we already have in `mainConstants.js`.

Not just that, but in TS the actual literal string value of an enum does not equal a string literal with that same value -  `ServerRoles.Admin !== 'server:admin'` in the eyes of TypeScript. So if you predict wanting to use a string literal from outside the enum (e.g. `Roles` imported from `mainConstants.js`), you should avoid typing the constant values with enums.

I've solved both of those problems by converting `mainConstants.js` to TS, and extracting ` ServerRoles` and `StreamRoles` types from the already defined constants there. Instead of being enum values, they will just be a union type of all of the allowed role values. This results in a less duplicated codebase and also in being able to use the `Roles` constant values wherever you expect `StreamRoles` or `ServerRoles`.

(I've pushed these changes into your branch)

<img width="600" alt="image" src="https://user-images.githubusercontent.com/938316/183933793-fb66ca00-9ccb-4c5e-9f7f-de7b7a993ae0.png">

Reply:
example of the ServerRoles.admin !== 'server:admin` issue:
<img width="464" alt="image" src="https://user-images.githubusercontent.com/938316/183939685-76804a15-b7e7-46ef-98d2-82d8507cb882.png">

Reply:
there was an inverse solution btw:
1. define ServerRoles & StreamRoles as enums
2. update the `Roles` constant and change all of its string literal values to enum values

this way you won't get TS errors when using a value off `Roles` in a variable/param that expects the `StreamRole` or `ServerRole` enum. if you would ever use a string literal directly, however, it wouldn't work